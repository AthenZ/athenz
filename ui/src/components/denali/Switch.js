/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { css, cx } from 'emotion';
import { rgba } from 'polished';
import { colors } from './styles/colors';
import { cssFontStyles } from './styles/fonts';

const makeCssSwitch = (props) => css`
    align-items: center;
    backface-visibility: hidden;
    box-sizing: border-box;
    display: inline-flex;
    height: 18px;
    outline: 0;
    position: relative;
    label: switch;

    & > input[type='checkbox'] {
        box-sizing: border-box;
        cursor: pointer;
        position: absolute; /* Take it out of the document flow */
        top: 0;
        visibility: hidden; /* Hide it */
    }

    & > input[type='checkbox'] + label {
        box-sizing: border-box;
        cursor: pointer;
        ${cssFontStyles.default};
        height: 18px;
        line-height: 18px;
        white-space: nowrap;

        & .label-content {
            display: inline-flex;
            margin-left: 6px;

            /* Firefox Quantum's text is off by 1px compared to Chrome/Safari,
       * so hack it to align with the other browsers. */
            @-moz-document url-prefix() {
                margin-top: 1px;
            }
        }
    }

    /* Line behind circle */
    & > input[type='checkbox'] + label:before {
        background: ${colors.grey500};
        border-radius: 9px;
        box-sizing: border-box;
        content: '';
        cursor: pointer;
        display: inline-block;
        height: 4px;
        line-height: 16px;
        margin: 7px 0 0 0;
        transition: all 0.25s;
        vertical-align: top;
        width: 34px;
    }

    /* Circle */
    & > input[type='checkbox'] + label::after {
        background-color: ${colors.white};
        border: 2px solid ${rgba(colors.brand600, 0.5)};
        border-radius: 50%;
        box-sizing: border-box;
        content: ' ';
        cursor: pointer;
        display: block;
        height: 18px;
        left: 0;
        position: absolute;
        transition: ${!props.noanim ? 'all 0.25s ease' : undefined};
        top: 0; /* Slight adjustment to due vertical-align: text-bottom */
        width: 18px;
    }

    /* Checked line behind circle */
    & > input[type='checkbox']:checked + label::before {
        background-color: ${rgba(colors.statusSuccess, 0.45)};
    }

    /* Checked circle */
    & > input[type='checkbox']:checked + label::after {
        background-color: ${colors.statusSuccess};
        border-color: ${colors.statusSuccess};
        left: 16px;
    }

    /* Disabled label */
    & > input[type='checkbox']:disabled + label {
        color: ${colors.grey600};
        cursor: not-allowed;
    }

    /* Disabled line behind circle */
    & > input[type='checkbox']:disabled + label::before {
        background-color: ${colors.grey400};
        cursor: not-allowed;
    }

    /* Disabled (unchecked) circle */
    & > input[type='checkbox']:disabled + label::after {
        border-color: ${colors.grey500};
        cursor: not-allowed;
    }

    /* Disabled checked circle */
    & > input[type='checkbox']:checked:disabled + label::after {
        background: ${colors.grey500};
    }
`;

/**
 * A wrapper around `<input type="checkbox>`, similar to `<Checkbox>`, except
 * this component is styled as a toggle switch instead.
 */
class Switch extends React.PureComponent {
    get switchId() {
        return this.props.id || `switch-${this.props.name.replace(/\s/g, '_')}`;
    }

    get label() {
        if (this.props.checked && this.props.labelOn) {
            return this.props.labelOn;
        } else if (!this.props.checked && this.props.labelOff) {
            return this.props.labelOff;
        }
        return this.props.label;
    }

    render() {
        /* eslint-disable no-unused-vars */
        const {
            className,
            innerRef,
            label,
            labelOn,
            labelOff,
            noanim,
            ...rest
        } = this.props;
        const cssProps = { noanim };
        const cssSwitch = makeCssSwitch(cssProps);

        const classes = cx(cssSwitch, 'denali-switch', className);

        return (
            <div className={classes} data-testid='switch-wrapper'>
                <input {...rest} id={this.switchId} ref={innerRef} />
                <label htmlFor={this.switchId}>
                    {this.label && (
                        <div className='label-content'>{this.label}</div>
                    )}
                </label>
            </div>
        );
    }
}

Switch.propTypes = {
    /** true = checked, false = unchecked */
    checked: PropTypes.bool.isRequired,
    /** Additonal class to apply to the outer div */
    className: PropTypes.string,
    /** ID to attach to the `<input>` element */
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    /** Forward React ref to the `<input>` element */
    innerRef: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
    /** Text to appear to the right of the checkbox */
    label: PropTypes.string,
    /** Used in conjunction with `labelOn` for seperate on/off labels (overrides `label`) */
    labelOff: PropTypes.string,
    /** Used in conjunction with `labelOff` for seperate on/off labels (overrides `label`) */
    labelOn: PropTypes.string,
    /**
     * Validator to ensure `labelOn` & `labelOff` are both defined or both undefined
     * @ignore
     */
    labelOnAndOff: function(props, propName, componentName) {
        if (props.labelOff && !props.labelOn) {
            return new Error(
                'Prop `labelOff` requires `labelOn` to also be defined in `' +
                    componentName +
                    '`. Validation failed.'
            );
        }
        if (props.labelOn && !props.labelOff) {
            return new Error(
                'Prop `labelOn` requires `labelOff` to also be defined in `' +
                    componentName +
                    '`. Validation failed.'
            );
        }
    },
    /** Name of the `<input>` element */
    name: PropTypes.string.isRequired,
    /** Disable animations / transitions */
    noanim: PropTypes.bool,
    /** @ignore */
    type: PropTypes.oneOf(['checkbox']),
    /** Value of the `<input>` element */
    value: PropTypes.oneOfType([
        PropTypes.bool,
        PropTypes.string,
        PropTypes.number,
    ]),
};

Switch.defaultProps = {
    noanim: false,
    type: 'checkbox',
};

export default Switch;
