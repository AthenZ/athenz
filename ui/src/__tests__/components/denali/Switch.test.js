/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import Switch from '../../../components/denali/Switch';

describe('Switch', () => {
    const ORIG_NODE_ENV = process.env.NODE_ENV;
    const onChange = jest.fn();

    afterEach(() => {
        onChange.mockClear();
        process.env.NODE_ENV = ORIG_NODE_ENV;
    });

    it('should render unchecked state', () => {
        const { getByTestId } = render(
            <Switch
                checked={false}
                label='test-label'
                name='test-switch'
                value='1'
                onChange={onChange}
            />
        );
        const wrapper = getByTestId('switch-wrapper');
        const inputNode = wrapper.querySelector('input');

        expect(wrapper).toMatchSnapshot();
        expect(inputNode).not.toHaveAttribute('checked');
        expect(inputNode).toHaveAttribute('id', 'switch-test-switch');
        expect(inputNode).not.toBeDisabled();
    });

    it('should render checked state', () => {
        const { getByTestId } = render(
            <Switch
                checked={true}
                label='test-label'
                name='test-switch'
                value='1'
                onChange={onChange}
            />
        );
        const wrapper = getByTestId('switch-wrapper');
        const inputNode = wrapper.querySelector('input');

        expect(wrapper).toMatchSnapshot();
        expect(inputNode).toHaveAttribute('checked');
        expect(inputNode).not.toBeDisabled();
    });

    it('should render disabled state', () => {
        const { getByTestId } = render(
            <Switch
                checked={false}
                disabled
                label='test-label'
                name='test-switch'
                value='1'
                onChange={onChange}
            />
        );
        const wrapper = getByTestId('switch-wrapper');
        const inputNode = wrapper.querySelector('input');

        expect(inputNode).toBeDisabled();
    });

    it('should handle id prop', () => {
        const { getByTestId } = render(
            <Switch
                checked={false}
                id='custom-id'
                label='test-label'
                name='test-switch'
                value='1'
                onChange={onChange}
            />
        );
        const wrapper = getByTestId('switch-wrapper');
        const inputNode = wrapper.querySelector('input');

        expect(inputNode).toHaveAttribute('id', 'custom-id');
    });

    it('should render with additional className', () => {
        const { getByTestId } = render(
            <Switch
                checked={false}
                className='custom-class'
                label='test-label'
                name='test-switch'
                value='1'
                onChange={onChange}
            />
        );
        const wrapper = getByTestId('switch-wrapper');

        expect(wrapper).toHaveClass('denali-switch');
        expect(wrapper).toHaveClass('custom-class');
    });

    it('should work with labelOn & labelOff', () => {
        const { queryByText, rerender } = render(
            <Switch
                checked={false}
                labelOn='test-label on'
                labelOff='test-label off'
                name='test-switch'
                value='1'
                onChange={onChange}
            />
        );

        expect(queryByText('test-label off')).not.toBeNull();
        expect(queryByText('test-label on')).toBeNull();

        rerender(
            <Switch
                checked={true}
                labelOn='test-label on'
                labelOff='test-label off'
                name='test-switch'
                value='1'
                onChange={onChange}
            />
        );

        expect(queryByText('test-label off')).toBeNull();
        expect(queryByText('test-label on')).not.toBeNull();
    });

    it('should warn on invalid labelOn & labelOff usage', () => {
        process.env.NODE_ENV = 'test';
        const consoleErrorSpy = jest
            .spyOn(console, 'error')
            .mockImplementation(() => {});

        render(
            <Switch
                checked
                name='test-switch1'
                labelOn='test-label on'
                value='1'
                onChange={onChange}
            />
        );
        render(
            <Switch
                checked
                name='test-switch2'
                labelOff='test-label off'
                value='1'
                onChange={onChange}
            />
        );

        expect(consoleErrorSpy).toHaveBeenCalledTimes(2);
    });

    it('should handle onChange event', () => {
        const { getByTestId } = render(
            <Switch
                checked={false}
                label='test-label'
                name='test-switch'
                value='1'
                onChange={onChange}
            />
        );
        const wrapper = getByTestId('switch-wrapper');
        const inputNode = wrapper.querySelector('input');

        fireEvent.click(inputNode);

        expect(onChange).toHaveBeenCalled();
    });
});
