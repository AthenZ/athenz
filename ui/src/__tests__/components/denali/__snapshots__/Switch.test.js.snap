// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should render checked state 1`] = `
.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 18px;
  outline: 0;
  position: relative;
}

.emotion-0 > input[type='checkbox'] {
  box-sizing: border-box;
  cursor: pointer;
  position: absolute;
  top: 0;
  visibility: hidden;
}

.emotion-0 > input[type='checkbox'] + label {
  box-sizing: border-box;
  cursor: pointer;
  font-family: Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 300;
  height: 18px;
  line-height: 18px;
  white-space: nowrap;
}

.emotion-0 > input[type='checkbox'] + label .label-content {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  margin-left: 6px;
}

@-moz-document url-prefix()  {
  .emotion-0 > input[type='checkbox'] + label .label-content {
    margin-top: 1px;
  }
}

.emotion-0 > input[type='checkbox'] + label:before {
  background: #d5d5d5;
  border-radius: 9px;
  box-sizing: border-box;
  content: '';
  cursor: pointer;
  display: inline-block;
  height: 4px;
  line-height: 16px;
  margin: 7px 0 0 0;
  -webkit-transition: all 0.25s;
  transition: all 0.25s;
  vertical-align: top;
  width: 34px;
}

.emotion-0 > input[type='checkbox'] + label::after {
  background-color: #fff;
  border: 2px solid rgba(54,151,242,0.5);
  border-radius: 50%;
  box-sizing: border-box;
  content: ' ';
  cursor: pointer;
  display: block;
  height: 18px;
  left: 0;
  position: absolute;
  -webkit-transition: all 0.25s ease;
  transition: all 0.25s ease;
  top: 0;
  width: 18px;
}

.emotion-0 > input[type='checkbox']:checked + label::before {
  background-color: rgba(21,192,70,0.45);
}

.emotion-0 > input[type='checkbox']:checked + label::after {
  background-color: #15c046;
  border-color: #15c046;
  left: 16px;
}

.emotion-0 > input[type='checkbox']:disabled + label {
  color: #9a9a9a;
  cursor: not-allowed;
}

.emotion-0 > input[type='checkbox']:disabled + label::before {
  background-color: #e8e8e8;
  cursor: not-allowed;
}

.emotion-0 > input[type='checkbox']:disabled + label::after {
  border-color: #d5d5d5;
  cursor: not-allowed;
}

.emotion-0 > input[type='checkbox']:checked:disabled + label::after {
  background: #d5d5d5;
}

<div
  class="emotion-0 denali-switch"
  data-testid="switch-wrapper"
>
  <input
    checked=""
    id="switch-test-switch"
    name="test-switch"
    type="checkbox"
    value="1"
  />
  <label
    for="switch-test-switch"
  >
    <div
      class="label-content"
    >
      test-label
    </div>
  </label>
</div>
`;

exports[`Switch should render unchecked state 1`] = `
.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 18px;
  outline: 0;
  position: relative;
}

.emotion-0 > input[type='checkbox'] {
  box-sizing: border-box;
  cursor: pointer;
  position: absolute;
  top: 0;
  visibility: hidden;
}

.emotion-0 > input[type='checkbox'] + label {
  box-sizing: border-box;
  cursor: pointer;
  font-family: Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 300;
  height: 18px;
  line-height: 18px;
  white-space: nowrap;
}

.emotion-0 > input[type='checkbox'] + label .label-content {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  margin-left: 6px;
}

@-moz-document url-prefix()  {
  .emotion-0 > input[type='checkbox'] + label .label-content {
    margin-top: 1px;
  }
}

.emotion-0 > input[type='checkbox'] + label:before {
  background: #d5d5d5;
  border-radius: 9px;
  box-sizing: border-box;
  content: '';
  cursor: pointer;
  display: inline-block;
  height: 4px;
  line-height: 16px;
  margin: 7px 0 0 0;
  -webkit-transition: all 0.25s;
  transition: all 0.25s;
  vertical-align: top;
  width: 34px;
}

.emotion-0 > input[type='checkbox'] + label::after {
  background-color: #fff;
  border: 2px solid rgba(54,151,242,0.5);
  border-radius: 50%;
  box-sizing: border-box;
  content: ' ';
  cursor: pointer;
  display: block;
  height: 18px;
  left: 0;
  position: absolute;
  -webkit-transition: all 0.25s ease;
  transition: all 0.25s ease;
  top: 0;
  width: 18px;
}

.emotion-0 > input[type='checkbox']:checked + label::before {
  background-color: rgba(21,192,70,0.45);
}

.emotion-0 > input[type='checkbox']:checked + label::after {
  background-color: #15c046;
  border-color: #15c046;
  left: 16px;
}

.emotion-0 > input[type='checkbox']:disabled + label {
  color: #9a9a9a;
  cursor: not-allowed;
}

.emotion-0 > input[type='checkbox']:disabled + label::before {
  background-color: #e8e8e8;
  cursor: not-allowed;
}

.emotion-0 > input[type='checkbox']:disabled + label::after {
  border-color: #d5d5d5;
  cursor: not-allowed;
}

.emotion-0 > input[type='checkbox']:checked:disabled + label::after {
  background: #d5d5d5;
}

<div
  class="emotion-0 denali-switch"
  data-testid="switch-wrapper"
>
  <input
    id="switch-test-switch"
    name="test-switch"
    type="checkbox"
    value="1"
  />
  <label
    for="switch-test-switch"
  >
    <div
      class="label-content"
    >
      test-label
    </div>
  </label>
</div>
`;
